### PR TITLE
Upgrade to Ruby 3.2

### PR DESCRIPTION
**Pull request by tenderlove**
This PR is based off #22585, but upgrades us to Ruby 3.2 instead of 3.1.

According to [the Docker image source](https://github.com/moritzheiber/ruby-jemalloc-docker), this build of Ruby 3.2 includes support for YJIT.

### CODE REVIEW

**Comment by shleeable** [comment]
Thank you very much for this.

I'd be interested to see tests using both `3.2.0-ubuntu-20.04` and `3.2.0-ubuntu-22.04`

**Comment by tenderlove** [comment]
> Thank you very much for this... Same as the 3.1 release, I'd be interested to see tests using both `3.2.0-ubuntu-20.04` and `3.2.0-ubuntu-22.04`

Ok, I'll try to figure out how to make sure both of those are in the build.  Thanks for the feedback! 😃

**Comment by nightpool** [comment]
Looks like there's still one test failure in outboxes spec:

```
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:95 # ActivityPub::OutboxesController GET #show without signature with page requested returns orderedItems with public or unlisted statuses
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:103 # ActivityPub::OutboxesController GET #show without signature with page requested returns Vary header with Signature
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:91 # ActivityPub::OutboxesController GET #show without signature with page requested returns application/activity+json
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:87 # ActivityPub::OutboxesController GET #show without signature with page requested returns http success
rspec './spec/controllers/activitypub/outboxes_controller_spec.rb[1:1:1:2:4:2]' # ActivityPub::OutboxesController GET #show without signature with page requested behaves like cacheable response does not set sessions
rspec './spec/controllers/activitypub/outboxes_controller_spec.rb[1:1:1:2:4:3]' # ActivityPub::OutboxesController GET #show without signature with page requested behaves like cacheable response returns public Cache-Control header
rspec './spec/controllers/activitypub/outboxes_controller_spec.rb[1:1:1:2:4:1]' # ActivityPub::OutboxesController GET #show without signature with page requested behaves like cacheable response does not set cookies
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:143 # ActivityPub::OutboxesController GET #show with signature when signed request account does not follow account returns application/activity+json
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:154 # ActivityPub::OutboxesController GET #show with signature when signed request account does not follow account returns private Cache-Control header
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:147 # ActivityPub::OutboxesController GET #show with signature when signed request account does not follow account returns orderedItems with public or unlisted statuses
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:139 # ActivityPub::OutboxesController GET #show with signature when signed request account does not follow account returns http success
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:169 # ActivityPub::OutboxesController GET #show with signature when signed request account follows account returns application/activity+json
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:173 # ActivityPub::OutboxesController GET #show with signature when signed request account follows account returns orderedItems with private statuses
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:180 # ActivityPub::OutboxesController GET #show with signature when signed request account follows account returns private Cache-Control header
rspec ./spec/controllers/activitypub/outboxes_controller_spec.rb:165 # ActivityPub::OutboxesController GET #show with signature when signed request account follows account returns http success
```

```
  1) ActivityPub::OutboxesController GET #show without signature with page requested returns orderedItems with public or unlisted statuses
     Failure/Error: Rails.cache.write(item, item)
     
     NoMethodError:
       undefined method `fetch_value' for nil:NilClass
     # ./app/controllers/concerns/cache_concern.rb:46:in `block in cache_collection'
     # ./app/controllers/concerns/cache_concern.rb:45:in `each_value'
     # ./app/controllers/concerns/cache_concern.rb:45:in `cache_collection'
     # ./app/controllers/concerns/cache_concern.rb:54:in `cache_collection_paginated_by_id'
     # ./app/controllers/activitypub/outboxes_controller.rb:64:in `set_statuses'
     # ./spec/controllers/activitypub/outboxes_controller_spec.rb:38:in `block (4 levels) in <top (required)>'
     # ./spec/spec_helper.rb:54:in `body_as_json'
     # ./spec/controllers/activitypub/outboxes_controller_spec.rb:39:in `block (4 levels) in <top (required)>'
     # ./spec/controllers/activitypub/outboxes_controller_spec.rb:96:in `block (5 levels) in <top (required)>'

  2) ActivityPub::OutboxesController GET #show without signature with page requested returns Vary header with Signature
     Failure/Error: Rails.cache.write(item, item)
     
     NoMethodError:
       undefined method `fetch_value' for nil:NilClass
     # ./app/controllers/concerns/cache_concern.rb:46:in `block in cache_collection'
     # ./app/controllers/concerns/cache_concern.rb:45:in `each_value'
     # ./app/controllers/concerns/cache_concern.rb:45:in `cache_collection'
     # ./app/controllers/concerns/cache_concern.rb:54:in `cache_collection_paginated_by_id'
     # ./app/controllers/activitypub/outboxes_controller.rb:64:in `set_statuses'
     # ./spec/controllers/activitypub/outboxes_controller_spec.rb:38:in `block (4 levels) in <top (required)>'
     # ./spec/controllers/activitypub/outboxes_controller_spec.rb:104:in `block (5 levels) in <top (required)>'
      # ./spec/controllers/activitypub/outboxes_controller_spec.rb:162:in `block (5 levels) in <top (required)>
```

All related to Rails.cache.write. Maybe something funky in the new edge version of ox?

File: Gemfile
Comment (p8) [inline Gemfile:68]: Isn't this just 2.14.12?
https://github.com/ohler55/ox/commit/cb8f1af3d2191178b413185ce8874244bbb920e5

Comment (tenderlove) [reply Gemfile:68]: @p8 thanks!  I'll update the Gemfile to point at the released version.

**Comment by shleeable** [comment]
Hopefully the next nokogiri update should come in with dependabot soon

**Comment by tenderlove** [comment]
> Hopefully the next nokogiri update should come in with dependabot soon

I'll look out for it and rebase.  Also I think maybe I'll close the Ruby 3.1 PR since we may as well target 3.2

**Comment by nightpool** [comment]
I would prefer to put the change into this PR if it's Ruby 3.2 related, no? Why wait for the unrelated dependabot PR to be merged if we know we're going to need to update anyway?

**Comment by shleeable** [comment]
@tenderlove can you give this a retry?

**Comment by tenderlove** [comment]
I'm not sure why this is failing in CI.  It looks like the cache store is null.  I can't seem to reproduce this locally, so I'm going to temporarily disable the backtrace cleaner in CI and see if that gives us better info.

**Comment by tenderlove** [comment]
Also I'll put this back to Draft until we can get a green CI.

**Comment by tenderlove** [comment]
Seems like the build might be flaky.  The CircleCI build was red in [9a226c1](https://github.com/mastodon/mastodon/pull/22928/commits/9a226c1eed3dd83b377a79f759ad92c50ca7fc4f), but green in  [5e14f47](https://github.com/mastodon/mastodon/pull/22928/commits/5e14f472b48ff6bba82e9cc83b9180ea5b8e7b4e).  I'm not sure what the deal is, so I'm just going to revert the debugging code I added.  I suspect maybe Redis isn't always available?

**Comment by renchap** [comment]
@tenderlove Thanks a lot for you work! Now that 4.1.0 is released, we are ready to merge this PR once it is ready (and then use it on some pods for one of our big instances to get an idea of the improvements with YJIT).

@Gargron released a new version of the `blurhash` gem, so it you can revert the github origin.

**Comment by shleeable** [comment]
+ Ruby released 3.2.1

**Comment by tenderlove** [comment]
> @Gargron released a new version of the `blurhash` gem, so it you can revert the github origin.

Thanks, I removed that.

> * Ruby released 3.2.1

Yep, I'll upgrade this PR when the Docker image is available! (I saw https://github.com/moritzheiber/ruby-jemalloc-docker/issues/3  😄)

**Review by ClearlyClaire — APPROVED** [review]

**Comment by moritzheiber** [comment]
@tenderlove Once [this pipeline](https://github.com/moritzheiber/ruby-jemalloc-docker/actions/runs/3939553433) has run successfully the container image(s) should be available 🙂

**Comment by tenderlove** [comment]
Ok, I've updated this to use the 3.2.1 docker container and removed a couple unnecessary commits.

Thanks!

File: Gemfile
Comment (ClearlyClaire) [inline Gemfile:64]: Mail *has* been updated to 2.8.0

File: Gemfile
Comment (renchap) [inline Gemfile]: Is there an issue opened to update to a more recent version of this gem?

This could be a temporary workaround but the latest version (2.x, so 2 major bumps to handle!) would be a much better solution (and will be needed for upgrading Rails as well)

File: Gemfile
Comment (tenderlove) [reply Gemfile]: Thanks, I didn't catch that.  It was not upgraded at the time I'd made the commits. 🙇 

I've removed the commit that added this.

File: Gemfile
Comment (tenderlove) [reply Gemfile]: There _was_ an issue, and I'll try to find it as a reference.  IIRC the problem was that the new version is very *not* backwards compatible, and the effort to upgrade was too much.

Comment (tenderlove) [reply Gemfile]: I couldn't find an issue, only [this comment](https://github.com/mastodon/mastodon/pull/21061#issuecomment-1341779139) and [this comment](https://github.com/mastodon/mastodon/pull/22585#discussion_r1057907260).

I'll file an issue so that we can track it.

Comment (tenderlove) [reply Gemfile]: I made an issue here: https://github.com/mastodon/mastodon/issues/23599

I filed it as a feature request because it didn't seem to fit the "Bug" template.  I was tempted to put it in discussions, but technical debt didn't really seem to fit in there either.

If you need me to change the location (or close it), please let me know.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by tenderlove** [comment]
Yay!! Thank you!!

**Comment by tisba** [comment]
Is mastodon using `bootsnap`? I run into an issue with Ruby 3.2, see https://github.com/Shopify/bootsnap/issues/436 and https://bugs.ruby-lang.org/issues/19419

**Comment by byroot** [comment]
Bootsnap ISeq compilation is disabled, so no concern here: https://github.com/mastodon/mastodon/blob/68a92c81bfb1bc4703cd4974321b8cb168793541/config/boot.rb#L15

**Pull request merged**
